Proper handling of CJK and accent characters

Mehrad0711 opened the pull request.

Mehrad0711 committed:
Proper handling of CJK characters

BERT tokenizes cjk characters by inserting whitespace between each char, whereas sentencepiece models consume the input as is.
During preprocessing we remove spaces between cjk chars (if any) and then let the numericalizer decide how to preprocess

Mehrad0711 committed:
Update XLMR tokenizer

Account for word-pieces not present in spm vocabulary.
Otherwise errors will be thrown when parser tries to generate those tokens.

gcampax reviewed (changes requested):
You cannot strip the accents, because the decoded words won't match the gold label any more when we evaluate in Genie (which does not strip the accents)

Mehrad0711 committed:
Update BERT tokenizer

Account for word-pieces not present in bert vocabulary.
Bert tokenizer strips accents. However we decide to extend vocab to include tokens with accents and other tokens not present in bert vocab.
This eliminates UNK issue we were encountering before.

gcampax commented on genienlp/data_utils/numericalizer/masked_tokenizer.py:
```diff
@@ -124,11 +130,31 @@ def tokenize(self, tokens, mask):
+                    token_id = len(self.vocab) + len(self.added_tokens_encoder)
+                    self.added_tokens_encoder[token] = token_id
+                    self.added_tokens_decoder[token_id] = token
+                output_tokens.append(token)
```
Is this adding the whole word or just the accent as a single `##` token? It would be nice not to lose the existing BERT embeddings around words with accents.

Mehrad0711 replied:
addressed.

gcampax reviewed (commented):
Please do a full round of training the Thingpedia model with this before merging it.
